File: dijkstra-algorithm/dijkstra_algorithm.js
function dijkstra(graph, start) {
    const distances = {};
    const pq = [[0, start]];
    
    for (let node in graph) distances[node] = Infinity;
    distances[start] = 0;
    
    while (pq.length) {
        pq.sort((a, b) => a[0] - b[0]);
        const [currentDistance, current] = pq.shift();
        
        if (currentDistance > distances[current]) continue;
        
        for (let neighbor in graph[current]) {
            const distance = currentDistance + graph[current][neighbor];
            if (distance < distances[neighbor]) {
                distances[neighbor] = distance;
                pq.push([distance, neighbor]);
            }
        }
    }
    return distances;
}

const graph = {A: {B: 1, C: 4}, B: {C: 2, D: 5}, C: {D: 1}, D: {}};
console.log(dijkstra(graph, 'A'));

File: dijkstra-algorithm/dijkstra_algorithm.rb
def dijkstra(graph, start)
  distances = Hash.new(Float::INFINITY)
  distances[start] = 0
  pq = [[0, start]]
  
  while !pq.empty?
    pq.sort!
    current_distance, current = pq.shift
    next if current_distance > distances[current]
    
    graph[current].each do |neighbor, weight|
      distance = current_distance + weight
      if distance < distances[neighbor]
        distances[neighbor] = distance
        pq.push([distance, neighbor])
      end
    end
  end
  
  distances
end

graph = {'A' => {'B' => 1, 'C' => 4}, 'B' => {'C' => 2, 'D' => 5}, 'C' => {'D' => 1}, 'D' => {}}
puts dijkstra(graph, 'A').inspect